---
title: "glmSparseNet"
output: 
  rmarkdown::html_vignette:
    toc: TRUE
  github_document:
    toc: TRUE
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.width = 6,
                      fig.height = 8)

library(futile.logger)
library(ggplot2)

.Last.value <- flog.layout(layout.format('[~l] ~m'))
.Last.value <- flog.appender(appender.tee('logger.txt'))
theme_set(theme_minimal())
```

```{r generate_md, eval=FALSE, include=FALSE}
rmarkdown::render('README.Rmd', output_format = 'all')
```

> Elastic-Net models with additional regularization based on network centrality metrics

## Overview

`glmSparseNet` is a R package that generalizes sparse regression models when the features have a graph structure (e.g. genes), by including network-based regularizers. `glmSparseNet` uses the glmnet package, by including centrality measures of the network as penality factors. The current version implements regularization based on node degree, i.e. the strength and/or number of its associated edges, either by promoting hubs in the solution (glmDegree) or orphan genes (glmOrphan) in the solution.
All the glmnet distribution families are supported, namely "gaussian", "poisson", "binomial", "multinomial", "cox", and "mgaussian".
Below, we provide one example for survival analysis of  .... using transcriptomic data of....RNA-seq tumor data.???  More information and RMD files are available in the folder ...???....where more extensive and complete examples are provided for logistic regressoin and... in ....



## Citation

Veríssimo, A., Oliveira, A.L., Sagot, M.-F., & Vinga, S. (2016). DegreeCox – a network-based regularization method for survival analysis. BMC Bioinformatics. 17(16): 449.  https://doi.org/10.1186/s12859-016-1310-4

A more detailed description of the extensions here developed will be released soon in a manuscript (under preparation).

This package was developed by André Veríssimo, Eunice Carrasquinha, Marta B. Lopes and Susana Vinga under the project SOUND, funded from the European Union Horizon 2020 research and innovation program under grant agreement No. 633974.



## Instalation

Bioconductor is necessary for the installation of this package.

```{r install, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite('averissimo/loose.rock')
biocLite('network.cox', siteRepos = 'https://sels.tecnico.ulisboa.pt/r-repos/')
```


To run the following examples, the next libraries are also needed:

```{r}
library(futile.logger)
library(dplyr)
library(ggplot2)
library(reshape2)
library(MultiAssayExperiment)
library(survival)
library(glmnet)
library(loose.rock)
library(network.cox)
```


This package extends the `glmnet` r-package with network-based regularization based on features relations. This network can be calculated from the data itself or using external networks to enrich the model.

It adds two new main functions called `network.glmnet` and `network.cv.glmnet` that extend both model inference and model selection via cross-validation with network-based regularization.

There are 3 methods available to use data-dependant methods to generate the network:

1. Correlation matrix with cutoff;
1. Covariance matrix with cutoff;
<!-- 1. Sparse bayesian networks using `sparsebn` package. -->

Alternatively, the network can be passed as an adjancency matrix or an already calculate metric for each node.

### Example for Gaussian models

The example below, shows random datasets being generated and `network.glmnet` new function being called.

```{r gaussian}
# Gaussian
x <- matrix(rnorm(100*20),100,20)
y <- rnorm(100)
fit1 <- network.glmnet(x,y, 'correlation', network.options = network.options.default(cutoff = 0.1))
```

Inspecting the penalty.factor used from correlation network.

```{r}
fit1$penalty.factor
```

Plot the results of the `glmnet` run.

```{r}
plot(fit1)
```

The given network parameter can also be a network itself, i.e. a matrix. The example below uses a randomly generated network to use in the methods.

```{r rand.network}
# generate random network
rand.network       <- abs(matrix(rnorm(20*20),20,20))
diag(rand.network) <- 0
# actual fit
fit4 <- network.glmnet(x,y, rand.network, network.options = network.options.default(cutoff = 0.1))
```

The result can be used with all functions available to glmnet objects, such as `predict`, `coef` or plot.

```{r predicted}
predicted <- predict(fit1, newx=x[1:10,],s=c(0.01,0.005))
```

```{r pred.table, echo=FALSE}
colnames(predicted) <- c('lambda_0.01', 'lambda_0.005')
flog.info('Observed vs. Predicted', cbind(Observed = y[1:10], predicted), capture = TRUE)
```

It also extends the new methods to the cross validation function with `network.cv.glmnet`.

```{r plot_cv}
plot(network.cv.glmnet(x,y, 'covariance'))
```

### Survival Example using RNASeq data

We use an example data from TCGA Adrenocortical Carcinoma project with '92' patients and a reduced RNASeq data. See `MultiAssayExperiment::miniACC` for more information and details of the data.

There is some pre-processing needed to remove patients with invalid follow-up date or death date:

```{r multi.assay}
library(MultiAssayExperiment)
# load data
xdata <- miniACC

# build valid data with days of last follow up or to event
event.ix <- which(!is.na(xdata$days_to_death))
cens.ix  <- which(!is.na(xdata$days_to_last_followup))
surv_event_time <- array(NA, nrow(xdata@colData))
surv_event_time[event.ix] <- xdata$days_to_death[event.ix]
surv_event_time[cens.ix]  <- xdata$days_to_last_followup[cens.ix]

# Keep only valid individuals
#
# they are valid if they have:
#  - either a follow_up time or event time
#  - a valid vital_status (i.e. not missing)
#  - folloup_time or event_time > 0
valid.ix <- as.vector(!is.na(surv_event_time) & !is.na(xdata$vital_status) & surv_event_time > 0)
ydata <- data.frame(time      = surv_event_time[valid.ix], 
                    status    = xdata$vital_status[valid.ix], 
                    row.names = xdata$patientID[valid.ix])

```

Fitting the survival model using a correlation network with cutoff at 0.6.

```{r fit.surv}
# build response object for glmnet
fit3 <- network.cv.glmnet(xdata, ydata, family = 'cox', 
                          network = 'correlation', 
                          experiment.name = 'RNASeq2GeneNorm', 
                          alpha = .7,
                          nlambda = 1000,
                          network.options = network.options.default(cutoff = .6, 
                                                                    min.degree = 0.2,
                                                                    trans.fun = degree.heuristic))
plot(fit3)
```

## Visualization tools

### Survival curves with `draw.kaplan`

This function generates Kaplan-Meier survival model based on the coefficients of a Cox model. It separates two groups based on relative risk and display both the curves and results of Log-rank test.

```{r draw.kaplan}
xdata.reduced <- reduce.by.experiment(xdata, 'RNASeq2GeneNorm')
ydata.km <- ydata[rownames(xdata.reduced@colData),]
best.model.coef <- coef(fit3, s = 'lambda.min')[,1]
draw.kaplan(best.model.coef, t(assay(xdata[['RNASeq2GeneNorm']])), ydata.km, ylim = c(0,1))
```

### Heatmap with results from Hallmarks of cancer

Search the non-zero coefficients in the results and query for known hallmarks of cancer.

```{r hallmarks}
hallmarks(names(best.model.coef)[best.model.coef > 0])$heatmap
```

```{r glmnet.classic, include=FALSE, eval=FALSE}
# Comparing against a classical GLMNET call
xdata.reduced <- filter.by.experiment(xdata, 'RNASeq2GeneNorm')
xdata.surv <- t(assay(xdata.reduced[['RNASeq2GeneNorm']]))
ydata.reduced <- ydata[xdata.reduced$patientID,]
ydata.surv <- Surv(ydata.reduced$time, ydata.reduced$status)

fit5 <- cv.glmnet(xdata.surv, ydata.surv, family = 'cox', alpha = .7, nlambda = 1000)
best.model.coef.g <- coef(fit5, s = 'lambda.min')[,1] %>% .[. != 0]


draw.kaplan(best.model.coef.g, 
            t(assay(xdata[['RNASeq2GeneNorm']]))[, names(best.model.coef.g)], 
            ydata.km,
            ylim = c(0,1))$plot

# hallmarks(names(best.model.coef.g))$heatmap
```

