---
title: "Example for Logistic regularization -- Breast Invasive Carcinoma"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
params:
  seed: !r 29221
  some_string: "some string"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 6,
  fig.height = 8
)
```

```{r, include=FALSE, eval=FALSE}
rmarkdown::render('example_brca_binomial.Rmd')
```


# Required Packages

```{r libraries}

library(reshape2)
library(tidyverse)
library(glmnet)
library(survival)
library(futile.logger)
library(loose.rock)
library(brca.data)

library(parallel)
library(doMC)

library(pROC)

# Load do package
devtools::load_all()

theme_set(theme_minimal())
.Last.value <- flog.layout(layout.format('[~l] ~m'))

.Last.value <- loose.rock::base.dir('./runcache')
.Last.value <- loose.rock::show.message(FALSE)
```

# Load data

Classification of Breast Invasive Carcinoma (BRCA) data from the Cancer Genome Atlas  (TCGA) ([TCGA](https://cancergenome.nih.gov/)).

```{r data}
library(brca.data)
data('fpkm.per.tissue', 'fpkm.per.tissue.barcode', 'clinical', package = 'brca.data')

# RNA-Seq data
# tumor tissue
xdata_tumor <- fpkm.per.tissue$primary.solid.tumor
xdata_tumor <- xdata_tumor[,-which(duplicated(getParticipantCode(colnames(fpkm.per.tissue$primary.solid.tumor))))] # removing duplicated samples
# normal tissue
set.seed(params$seed)
xdata_normal <- fpkm.per.tissue$solid.tissue.normal
xdata <- rbind(t(xdata_tumor),t(xdata_normal))

xdata <- xdata[, sample(which(apply(xdata, 2, sd) != 0), 300)]
# response vector ('1', tumor; '0', normal)
ydata <- c(rep(1,dim(xdata_tumor)[2]),rep(0,dim(xdata_normal)[2]))
```


# Fit models

```{r fit}
fitted <- run.cache(network.cv.glmnet, 
                    xdata, ydata, 
                    network = 'correlation',
                    family = 'binomial',
                    seed = params$seed)
```

# Results of Cross Validation

```{r results}
plot(fitted)
```

## Using lambda.min taken from cross validation

```{r top_coefs, echo=FALSE}
top.coefs <- coef(fitted, s = 'lambda.min')
coefs.v <- as.vector(top.coefs)
names(coefs.v) <- rownames(top.coefs)
flog.info('Non-zero coefficients:', coefs.v[coefs.v != 0], capture = T)
```

### Hallmarks of Cancer

```{r hallmarks, echo=FALSE}
names.df <- gene.names(names(coefs.v[coefs.v != 0]))
all.hallmarks <- hallmarks(names.df$external_gene_name)

all.hallmarks$hallmarks$gene.name <- rownames(all.hallmarks$hallmarks)

melt(all.hallmarks$hallmarks, 
                  id.vars = c('gene.name')) %>% filter(value > 0) %>%
  ggplot(aes(gene.name,variable, fill=value)) + 
    geom_raster() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
    ggtitle('Hallmarks heatmap', subtitle = sprintf('Genes without hallmarks (%d): %s', length(all.hallmarks$no.hallmakrs), paste(all.hallmarks$no.hallmakrs, collapse = ', '))) + 
    xlab('External Gene Name') + ylab('')
```

### Accuracy

```{r accuracy, echo=FALSE}
resp <- glmnet::predict.cv.glmnet(fitted, s = 'lambda.min', newx = xdata, type = 'response')
flog.info('Misclassified (%d)', sum(abs(ydata - resp) > 0.5))
flog.info('  * False positive: %d', sum(ydata - resp > 0.5))
flog.info('  * False negative: %d', sum(ydata - resp < -0.5))
```

```{r predict, echo=FALSE}
qplot(resp, bins = 100)
qplot(resp, ydata, color = abs(resp - ydata) < 0.5) + theme(legend.position = 'none')
```

```{r roc, echo=FALSE}
roc_obj <- roc(ydata, as.vector(resp))

data.frame(TPR = roc_obj$sensitivities, FPR = 1 - roc_obj$specificities) %>%
  ggplot() +geom_line(aes(FPR,TPR), color = 2, size = 1, alpha = 0.7)+
      labs(title= sprintf("ROC curve (AUC = %f)", auc(roc_obj)), 
           x = "False Positive Rate (1-Specificity)", 
           y = "True Positive Rate (Sensitivity)")
```

