---
title: "Example for Logistic regularization -- Breast Invasive Carcinoma"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
params:
  seed: !r 29221
  some_string: "some string"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 6,
  fig.height = 7
)

# Load do package
devtools::load_all()

ggplot2::theme_set(ggplot2::theme_minimal())
.Last.value <- futile.logger::flog.layout(futile.logger::layout.format('[~l] ~m'))

.Last.value <- loose.rock::base.dir(tempdir())
.Last.value <- loose.rock::show.message(FALSE)
```

# Required Packages

```{r libraries}
library(futile.logger)
library(ggplot2)
library(reshape2)
library(dplyr)
library(survival)
library(loose.rock)
library(glmnet)
```

```{r, eval=FALSE}
library(glmSparseNet)
```

# Load data

Classification of Breast Invasive Carcinoma (BRCA) data from the Cancer Genome Atlas  (TCGA) ([TCGA](https://cancergenome.nih.gov/)).

```{r data, eval=FALSE}
# download data package with TCGA data
biocLite('https://github.com/averissimo/tcga.data/releases/download/2016.12.15-brca/brca.data_1.0.tar.gz')

# load library
library(brca.data)

# load data
data('fpkm.per.tissue', 'fpkm.per.tissue.barcode', 'clinical', package = 'brca.data')

#
# RNA-Seq data

# tumor tissue
xdata_tumor <- fpkm.per.tissue$primary.solid.tumor
xdata_tumor <- xdata_tumor[,-which(duplicated(getParticipantCode(colnames(fpkm.per.tissue$primary.solid.tumor))))] # removing duplicated samples

# normal tissue
set.seed(params$seed)
xdata_normal <- fpkm.per.tissue$solid.tissue.normal
xdata        <- rbind(t(xdata_tumor),t(xdata_normal))
xdata        <- xdata[, apply(xdata, 2, sd) != 0] # remove genes with sd == 0

# response vector ('1', tumor; '0', normal)
ydata <- c(rep(1,dim(xdata_tumor)[2]),rep(0,dim(xdata_normal)[2]))

# Using only a subset of genes previously selected to keep this example short.

set.seed(params$seed)
small.subset <- c('ENSG00000110448', 'ENSG00000124449', 'ENSG00000103494', 
                  'ENSG00000172053', 'ENSG00000136758', 'ENSG00000051382', 
                  'ENSG00000128655', 'ENSG00000100368', 'ENSG00000115884', 
                  'ENSG00000179363', 'ENSG00000091106', 'ENSG00000178403', 
                  'ENSG00000066422', 'ENSG00000110448', 'ENSG00000124449', 
                  'ENSG00000230445', 'ENSG00000172053', 'ENSG00000185122', 
                   sample(colnames(xdata), 100))
xdata <- xdata[,small.subset]
```

```{r, eval=FALSE, include=FALSE}
sample.BRCA.logisitic <- list(xdata = xdata, ydata = ydata)
devtools::use_data(sample.BRCA.logistic, overwrite = TRUE)
```

```{r, include=FALSE}
xdata <- sample.BRCA.logistic$xdata
ydata <- sample.BRCA.logistic$ydata
```


# Fit models

```{r fit.show}
fitted <- network.cv.glmnet(xdata, ydata, network = 'correlation', family = 'binomial', nlambda = 1000)
```

# Results of Cross Validation

```{r results}
plot(fitted)
```

## Using lambda.min taken from cross validation

```{r top_coefs, echo=FALSE}
top.coefs <- coef(fitted, s = 'lambda.min')
coefs.v <- as.vector(top.coefs)
names(coefs.v) <- rownames(top.coefs)
flog.info('Non-zero coefficients:', coefs.v[coefs.v != 0], capture = T)
```

### Hallmarks of Cancer

```{r hallmarks, echo=FALSE}
names.df <- gene.names(names(coefs.v[coefs.v != 0]))
all.hallmarks <- hallmarks(names.df$external_gene_name)

all.hallmarks$heatmap
```

### Accuracy

```{r accuracy, echo=FALSE}
resp <- glmnet::predict.cv.glmnet(fitted, s = 'lambda.min', newx = xdata, type = 'response')
flog.info('Misclassified (%d)', sum(abs(ydata - resp) > 0.5))
flog.info('  * False positive: %d', sum(ydata - resp > 0.5))
flog.info('  * False negative: %d', sum(ydata - resp < -0.5))
```

Histogram of predicted response

```{r predict, echo=FALSE}
qplot(resp, bins = 100)
```

Prediction vs. Observed

```{r}
qplot(resp, ydata, color = abs(resp - ydata) < 0.5) + theme(legend.position = 'none')
```

ROC curve

```{r roc, echo=FALSE}
roc_obj <- pROC::roc(ydata, as.vector(resp))

data.frame(TPR = roc_obj$sensitivities, FPR = 1 - roc_obj$specificities) %>%
  ggplot() +geom_line(aes(FPR,TPR), color = 2, size = 1, alpha = 0.7)+
      labs(title= sprintf("ROC curve (AUC = %f)", pROC::auc(roc_obj)), 
           x = "False Positive Rate (1-Specificity)", 
           y = "True Positive Rate (Sensitivity)")
```

